iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -0,0 +1,216 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/LLVMGPUUtils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+constexpr unsigned kWorkgroupTileLevel = 0;
+constexpr int kSharedMemSizeBytes = 64 * 1024;
+
+LogicalResult verifyGPUMatmulSimtPassPipeline(
+    Operation *op, IREE::Codegen::LoweringConfigAttr loweringConfig,
+    IREE::Codegen::TranslationInfoAttr translationInfo,
+    ArrayRef<int64_t> workgroupSize) {
+  auto pipeline =
+      IREE::Codegen::DispatchLoweringPassPipeline::LLVMGPUMatmulSimt;
+  StringRef pipelineName = stringifyEnum(pipeline);
+  if (workgroupSize.empty()) {
+    return op->emitOpError("expected workgroup size for GPU pipelines");
+  }
+
+  if (!isa<linalg::MatmulOp, linalg::BatchMatmulOp>(op)) {
+    return success();  // Only verify batched and unbatched matmul.
+  }
+
+  Type inputType = op->getOperand(0).getType();
+  SmallVector<int64_t> firstLevelTileSizes =
+      loweringConfig.getTileSizeVals(kWorkgroupTileLevel);
+
+  if (linalg::BatchMatmulOp batchMatmulOp =
+          dyn_cast<linalg::BatchMatmulOp>(op)) {
+    // First tile dimensions should be 1 for batched, use remaining dimensions
+    // for comparisons.
+    if (firstLevelTileSizes[0] != 1) {
+      op->emitError("Received first tile dimension of ")
+          << firstLevelTileSizes[0] << " instead of 1 for " << pipelineName;
+    }
+    firstLevelTileSizes = {firstLevelTileSizes[1], firstLevelTileSizes[2],
+                           firstLevelTileSizes[3]};
+  }
+
+  // Verify the total workgroup size is <= 1024
+  int64_t totalWorkgroupSize =
+      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
+  if (totalWorkgroupSize > 1024) {
+    return op->emitOpError("expected workgroup size to be <=1024 for ")
+           << pipelineName << ", got " << totalWorkgroupSize;
+  }
+
+  // Verify the workgroup.z component should always be 1
+  if (workgroupSize[2] != 1) {
+    return op->emitOpError("expected workgroup z component to be 1 for ")
+           << pipelineName << ", got " << workgroupSize[2];
+  }
+
+  // Verify shared memory usage of operands after tiling requires <= 64Kb
+  // combined space.
+  unsigned bytesSize;
+  if (MemRefType type = inputType.dyn_cast<mlir::MemRefType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else if (RankedTensorType type = inputType.dyn_cast<RankedTensorType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else if (UnrankedTensorType type =
+                 inputType.dyn_cast<UnrankedTensorType>()) {
+    bytesSize = type.getElementType().getIntOrFloatBitWidth() / 8;
+  } else {
```

Comment:
nit: unranked tensor type shouldn't be possible in this case. I would remove it.